### PR TITLE
feature: Framework for Jobs to Target Runners by Label

### PR DESCRIPTION
**Pull request by xiaolin-ninja**
Part of Targetable Runners saga. This adds logic that matches jobs to runners by labels. It supports full and partial label matching.

A changelog will be included in the next PR that wires this up with user-facing changes to profile-based targeting.

### CODE REVIEW

**Comment by xiaolin-ninja** [comment]
go tests are failing because of check vendor test, but I have no changes to my `go.mod` compared to main. Please help. :'(

**Review by mitchellh — CHANGES_REQUESTED** [review]
Woohoo! Super solid. My only real request is for a few additional edge cases. I'm sure they pass but just want them in there for regression purposes in the future.

**Comment by mitchellh** [comment]
The failures are my fault, my suggestions on GH weren't formatted. Just need to run a `make format` locally.

File: internal/server/singleprocess/state/job.go [hunk 1]
```diff
@@ -1382,6 +1400,57 @@ func (s *State) jobCandidateById(
...
+			return nil, err
+		} else if blocked && assign {
+			continue
+		}
+
+		// Check whether job target labels match with runner labels
+		match := true
+		for k, v := range job.TargetRunnerLabels {
+			if val, ok := r.Runner.Labels[k]; !ok || v != val {
+				match = false
```
[the remark this answers is not in the document]
Comment (xiaolin-ninja) [reply internal/server/singleprocess/state/job.go:1443]: Wait, we would want to `break` out of the loop as soon as we hit a mismatch, and a continue outside of the loop to return `nil,nil` right? Otherwise a `continue` inside the loop would just go on to the next iteration?

[the remark this answers is not in the document]
Comment (mitchellh) [reply internal/server/singleprocess/state/job.go:1443]: Oh you're right. Nevermind keep this. 😄

**Review by mitchellh — APPROVED** [review]
This is safe to merge into the 0.8-series even if the full runner targeting by labels doesn't get done by 0.8 😄  (once everything is green)
